src/python/grpcio/grpc/__init__.py

```diff
@@ -27,4 +27,716 @@
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def code(self):
+    """Accesses the status code emitted by the service-side of the RPC.
+
+    This method blocks until the value is available.
+
+    Returns:
+      The integer status code of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def details(self):
+    """Accesses the details value emitted by the service-side of the RPC.
+
+    This method blocks until the value is available.
+
+    Returns:
+      The bytes of the details of the RPC.
+    """
+    raise NotImplementedError()
+
+
+class ServicerContext(six.with_metaclass(abc.ABCMeta, RpcContext)):
+  """A context object passed to method implementations."""
+
+  @abc.abstractmethod
+  def invocation_metadata(self):
+    """Accesses the metadata from the invocation-side of the RPC.
+
+    Returns:
+      The invocation metadata object as a sequence of pairs of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def initial_metadata(self, initial_metadata):
+    """Accepts the initial metadata value of the RPC.
+
+    This method need not be called by method implementations if they have no
+    service-side initial metadata to transmit.
+
+    Args:
+      initial_metadata: The initial metadata of the RPC as a sequence of pairs
+        of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def terminal_metadata(self, terminal_metadata):
+    """Accepts the terminal metadata value of the RPC.
+
+    This method need not be called by method implementations if they have no
+    service-side terminal metadata to transmit.
+
+    Args:
+      terminal_metadata: The terminal metadata of the RPC as a sequence of pairs
+        of bytes.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def code(self, code):
+    """Accepts the status code of the RPC.
+
+    This method need not be called by method implementations if they wish the
+    gRPC runtime to determine the status code of the RPC.
+
+    Args:
+      code: The integer status code of the RPC to be transmitted to the
+        invocation side of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def details(self, details):
+    """Accepts the service-side details of the RPC.
+
+    This method need not be called by method implementations if they have no
+    details to transmit.
+
+    Args:
+      details: The details bytes of the RPC to be transmitted to
+        the invocation side of the RPC.
+    """
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def peer(self):
+    """Identifies the peer that invoked the RPC being serviced.
+
+    Returns:
+      A string identifying the peer that invoked the RPC being serviced.
+    """
+    raise NotImplementedError()
+
+
+class UnaryUnaryMultiCallable(six.with_metaclass(abc.ABCMeta)):
+  """Affords invoking a unary-unary RPC."""
```

Comment:
Okay, wrapping this up: **I think we've heard quite loudly from our users that the `grpc` Python package is where they expect every developer-user-facing code element to be found**. I can think of two ways to go about satisfying that requirement:

1. put everything in `grpc.__init__.py`, as is currently drafted, or
2. spread code elements out reasonably into files like `_future.py`, `_multi_callables.py`, `_context.py`, and then alias them in `grpc.__init__.py`, being careful that the names by which they are officially known are the aliases in the `grpc` namespace and their having been defined in submodules is just an implementation artifact.

I favor the first choice, but I could live with the second, because both deliver the desired developer-user experience of
```
>>> import grpc
>>> help(grpc)

>>> help(grpc.Future)

>>> help(grpc.Channel)

>>> help(grpc.insecure_channel)
```
. What's _not_ part of the developer-user experience that we need to deliver is "the text of `grpc/__init__.py` has to be not too long. Developer-users should be learning gRPC Python from [the documentation](http://www.grpc.io/grpc/python/) or maybe from the read-eval-print loop as above.

The length of `grpc/__init__.py` is only a problem for us who maintain it and for power developer-users who want to change it. If it were an implementation module, I'd feel like it was on the long side, but since it's nearly all non-executable specification text I think it's just fine.

Please insist on breaking it up if you feel strongly, otherwise have I persuaded you at all?